gradle k8sBuild on a quarkus project is failing with java.util.ConcurrentModificationException

### Description
<!--
Thank you for trying out Eclipse JKube!

Please fill out the following information to help us in analyzing the issue, but feel free to skip it if you don't have the information at hand or if it does not apply. Please remove every section which does not apply to the issue.
-->

### Info

* Eclipse JKube version : 1.5.1 or 1.6.0-SNAPSHOT
* Gradle version: gradlew (provided) 7.3

* If it's a bug, how to reproduce :
1. create a new gradle Quarkus project from https://code.quarkus.io/
2. add to build.gradle
   ```
    id 'org.eclipse.jkube.kubernetes' version '1.5.1'
   ```
3. run 
   ```
   ./gradlew k8sBuild --info --stacktrace
   ```
4. it fails with 
   ```
    > Task :k8sBuild FAILED
    Caching disabled for task ':k8sBuild' because:
      Build cache is disabled
    Task ':k8sBuild' is not up-to-date because:
      Task has not declared any outputs despite executing actions.
    :k8sBuild (Thread[Execution worker for ':',5,main]) completed. Took 0.477 secs.

    FAILURE: Build failed with an exception.

    * What went wrong:
    Execution failed for task ':k8sBuild'.
    > java.util.ConcurrentModificationException (no error message)
   ```

```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':k8sBuild'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:145)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:143)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:402)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:389)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:382)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:368)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
Caused by: java.util.ConcurrentModificationException
        at org.gradle.api.internal.DefaultDomainObjectCollection$IteratorImpl.next(DefaultDomainObjectCollection.java:475)
        at org.eclipse.jkube.gradle.plugin.GradleUtil.extractDependencies(GradleUtil.java:138)
        at org.eclipse.jkube.gradle.plugin.GradleUtil.extractDependencies(GradleUtil.java:105)
        at org.eclipse.jkube.gradle.plugin.GradleUtil.convertGradleProject(GradleUtil.java:74)
        at org.eclipse.jkube.gradle.plugin.task.AbstractJKubeTask.runTask(AbstractJKubeTask.java:66)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@ Usage:
 ./scripts/extract-changelog-for-version.sh 1.3.37 5
 ```
 ### 1.6.0-SNAPSHOT
+* Fix #1136: ConcurrentModificationException when running gradle k8sBuild on Quarkus sample
 * Fix #887: Incorrect warning about overriding environment variable
 * Fix #802: Update Fabric8 kubernetes Client to v5.10.1
 * Fix #961: `k8sBuild`, `k8sResource` and `k8sApply` tasks don't respect skip options

--- a/gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/GradleUtil.java
+++ b/gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/GradleUtil.java
@@ -16,6 +16,7 @@ package org.eclipse.jkube.gradle.plugin;
 import java.io.File;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
@@ -111,7 +112,7 @@ public class GradleUtil {
 
   private static List<Dependency> extractDependencies(Project gradleProject,
       Function<ResolutionResult, Set<? extends DependencyResult>> resolutionToDependency) {
-    return gradleProject.getConfigurations().stream()
+    return new ArrayList<Configuration>(gradleProject.getConfigurations()).stream()
         .filter(GradleUtil::canBeResolved)
         .flatMap(c -> {
           final Map<ComponentIdentifier, ResolvedArtifactResult> artifacts = artifactMap(c);
@@ -179,7 +180,7 @@ public class GradleUtil {
   }
 
   private static File findArtifact(Project gradleProject) {
-    return gradleProject.getConfigurations().stream()
+    return new ArrayList<Configuration>(gradleProject.getConfigurations()).stream()
         .map(Configuration::getOutgoing)
         .map(ConfigurationPublications::getArtifacts)
         .map(PublishArtifactSet::getFiles)

--- a/gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/GradleUtilTest.java
+++ b/gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/GradleUtilTest.java
@@ -16,11 +16,14 @@ package org.eclipse.jkube.gradle.plugin;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
 import java.util.UUID;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -43,6 +46,7 @@ import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
+import org.mockito.stubbing.Answer;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.tuple;
@@ -61,12 +65,20 @@ public class GradleUtilTest {
   private Project project;
   private JavaPluginConvention javaPlugin;
 
+  private List<Configuration> projectConfigurations;
+
   @Before
   public void setUp() throws IOException {
     project = mock(Project.class, RETURNS_DEEP_STUBS);
     javaPlugin = mock(JavaPluginConvention.class, RETURNS_DEEP_STUBS);
     when(javaPlugin.getSourceSets().stream()).thenReturn(Stream.empty());
-    when(project.getConfigurations().stream()).thenAnswer(i -> Stream.empty());
+
+    final ConfigurationContainer cc = mock(ConfigurationContainer.class);
+    when(project.getConfigurations()).thenReturn(cc);
+    projectConfigurations = new ArrayList<Configuration>();
+    when(cc.stream()).thenAnswer(i -> projectConfigurations.stream());
+    when(cc.toArray()).thenAnswer(i -> projectConfigurations.toArray());
+
     when(project.getBuildscript().getConfigurations().stream()).thenAnswer(i -> Stream.empty());
     when(project.getProperties()).thenReturn(Collections.emptyMap());
     when(project.getBuildDir()).thenReturn(folder.newFolder("build"));
@@ -128,16 +140,13 @@ public class GradleUtilTest {
   @Test
   public void extractDependencies_withMultipleAndDuplicateDependencies_shouldReturnValidDependencies() {
     // Given
-    final ConfigurationContainer cc = mock(ConfigurationContainer.class);
-    when(project.getConfigurations()).thenReturn(cc);
     final Function<String[], Configuration> mockConfiguration = configurationDependencyMock();
-    when(cc.stream()).thenAnswer(i -> Stream.of(
-        mockConfiguration.apply(new String[] { "api", "com.example", "artifact", null }),
-        mockConfiguration.apply(new String[] { "implementation", "com.example.sub", "duplicate-artifact", "1.33.7" }),
-        mockConfiguration.apply(new String[] { "implementation", "com.example.sub", "duplicate-artifact", "1.33.7" }),
-        mockConfiguration.apply(new String[] { "implementation", "com.example", "other.artifact", "1.0.0" }),
-        mockConfiguration.apply(new String[] { "implementation", "com.example", "other.artifact", "1.1.0" }),
-        mockConfiguration.apply(new String[] { "testImplementation", "com.example", "other.artifact", "1.1.0" })));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "api", "com.example", "artifact", null }));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "implementation", "com.example.sub", "duplicate-artifact", "1.33.7" }));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "implementation", "com.example.sub", "duplicate-artifact", "1.33.7" }));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "implementation", "com.example", "other.artifact", "1.0.0" }));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "implementation", "com.example", "other.artifact", "1.1.0" }));
+    projectConfigurations.add(mockConfiguration.apply(new String[] { "testImplementation", "com.example", "other.artifact", "1.1.0" }));
     // When
     final JavaProject result = convertGradleProject(project);
     // Then
@@ -178,6 +187,32 @@ public class GradleUtilTest {
             tuple("org.springframework.boot", "org.springframework.boot.gradle.plugin", "1.33.7"),
             tuple("org.eclipse.jkube.kubernetes", "org.eclipse.jkube.kubernetes.gradle.plugin", "1.0.0")
         );
+  }
+
+ /**
+   * Some plugins might modify the ConfigurationContainer collection while we traverse it.
+   *
+   * This test ensures that no ConcurrentModificationException are thrown because we clone the ConfigurationContainer
+   * into an ArrayList.
+   */
+  @Test
+  public void convertGradleProject_withConcurrentConfigurationModifications_shouldReturnValidProject() {
+    // Given
+    final Configuration mockConfiguration = mock(Configuration.class, RETURNS_DEEP_STUBS);
+    projectConfigurations.add(mockConfiguration);
+    final Answer<Set<?>> concurrentModificationAnswer = i -> {
+      projectConfigurations.remove(mockConfiguration);
+      projectConfigurations.add(mockConfiguration);
+      return Collections.emptySet();
+    };
+    when(mockConfiguration.isCanBeResolved()).thenReturn(true);
+    when(mockConfiguration.getOutgoing().getArtifacts().getFiles().getFiles()).thenAnswer(concurrentModificationAnswer);
+    when(mockConfiguration.getIncoming().getResolutionResult().getAllDependencies()).thenAnswer(concurrentModificationAnswer);
+    when(mockConfiguration.getIncoming().getResolutionResult().getRoot().getDependencies()).thenAnswer(concurrentModificationAnswer);
+    // When
+    final JavaProject result = convertGradleProject(project);
+    // Then
+    assertThat(result).isNotNull();
   }
 
   @Test
@@ -221,7 +256,7 @@ public class GradleUtilTest {
     when(c.getOutgoing().getArtifacts().getFiles().getFiles()).thenReturn(Stream.of(
         folder.newFile("final-artifact.jar")
     ).collect(Collectors.toSet()));
-    when(project.getConfigurations().stream()).thenAnswer(i -> Stream.of(c));
+    projectConfigurations.add(c);
     // When
     final JavaProject result = convertGradleProject(project);
     // Then
@@ -239,7 +274,7 @@ public class GradleUtilTest {
     when(c.getOutgoing().getArtifacts().getFiles().getFiles()).thenReturn(Stream.of(
         jar2, jar1
     ).collect(Collectors.toSet()));
-    when(project.getConfigurations().stream()).thenAnswer(i -> Stream.of(c));
+    projectConfigurations.add(c);
     // When
     final JavaProject result = convertGradleProject(project);
     // Then

--- a/gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/TaskEnvironment.java
+++ b/gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/TaskEnvironment.java
@@ -16,6 +16,8 @@ package org.eclipse.jkube.gradle.plugin.task;
 import org.eclipse.jgit.util.FileUtils;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.plugins.JavaPluginConvention;
 import org.junit.rules.TemporaryFolder;
@@ -23,6 +25,8 @@ import org.mockito.MockedConstruction;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.stream.Stream;
 
 import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
@@ -43,7 +47,13 @@ public class TaskEnvironment extends TemporaryFolder {
     logger = mock(Logger.class);
     when(project.getProjectDir()).thenReturn(getRoot());
     when(project.getBuildDir()).thenReturn(newFolder("build"));
-    when(project.getConfigurations().stream()).thenAnswer(i -> Stream.empty());
+
+    final ConfigurationContainer cc = mock(ConfigurationContainer.class);
+    when(project.getConfigurations()).thenReturn(cc);
+    List<Configuration> projectConfigurations = new ArrayList<Configuration>();
+    when(cc.stream()).thenAnswer(i -> projectConfigurations.stream());
+    when(cc.toArray()).thenAnswer(i -> projectConfigurations.toArray());
+
     when(project.getBuildscript().getConfigurations().stream()).thenAnswer(i -> Stream.empty());
     when(project.getConvention().getPlugin(JavaPluginConvention.class)).thenReturn(mock(JavaPluginConvention.class));
     defaultTaskMockedConstruction = mockConstruction(DefaultTask.class, (mock, ctx) -> {

--- a/gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/task/TaskEnvironment.java
+++ b/gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/task/TaskEnvironment.java
@@ -15,11 +15,15 @@ package org.eclipse.jkube.gradle.plugin.task;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.stream.Stream;
 
 import org.eclipse.jgit.util.FileUtils;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.plugins.JavaPluginConvention;
 import org.junit.rules.TemporaryFolder;
@@ -43,7 +47,13 @@ public class TaskEnvironment extends TemporaryFolder {
     logger = mock(Logger.class);
     when(project.getProjectDir()).thenReturn(getRoot());
     when(project.getBuildDir()).thenReturn(newFolder("build"));
-    when(project.getConfigurations().stream()).thenAnswer(i -> Stream.empty());
+
+    final ConfigurationContainer cc = mock(ConfigurationContainer.class);
+    when(project.getConfigurations()).thenReturn(cc);
+    List<Configuration> projectConfigurations = new ArrayList<Configuration>();
+    when(cc.stream()).thenAnswer(i -> projectConfigurations.stream());
+    when(cc.toArray()).thenAnswer(i -> projectConfigurations.toArray());
+
     when(project.getBuildscript().getConfigurations().stream()).thenAnswer(i -> Stream.empty());
     when(project.getConvention().getPlugin(JavaPluginConvention.class)).thenReturn(mock(JavaPluginConvention.class));
     defaultTaskMockedConstruction = mockConstruction(DefaultTask.class, (mock, ctx) -> {

